eventstreamrpc/source/EventStreamClient.cpp

```diff
@@ -0,0 +1,1301 @@
+            {
+                if (m_messageCount == 1)
+                {
+                    RpcError promiseValue = {(EventStreamRpcStatusCode)errorCode, 0};
+                    m_initialResponsePromise.SetValue(TaggedResult(promiseValue));
+                }
+                else
+                {
+                    bool shouldClose = true;
+                    if (m_streamHandler)
+                        shouldClose = m_streamHandler->OnStreamError(nullptr, {errorCode, 0});
+                    if (!m_isClosed.load() && shouldClose)
+                    {
+                        Close().wait();
+                    }
+                }
+            }
+        }
+
+        std::future<RpcError> ClientOperation::Activate(
+            const OperationRequest *shape,
+            OnMessageFlushCallback onMessageFlushCallback) noexcept
+        {
+            /* Promises must be reset in case the client would like to send a subsequent request with the same
+             * `ClientOperation`. */
+            m_initialResponsePromise.Reset();
+            m_closedPromise = {};
+
+            Crt::List<EventStreamHeader> headers;
+            headers.emplace_back(EventStreamHeader(
+                Crt::String(CONTENT_TYPE_HEADER), Crt::String(CONTENT_TYPE_APPLICATION_JSON), m_allocator));
+            headers.emplace_back(
+                EventStreamHeader(Crt::String(SERVICE_MODEL_TYPE_HEADER), GetModelName(), m_allocator));
+            Crt::JsonObject payloadObject;
+            shape->SerializeToJsonObject(payloadObject);
+            Crt::String payloadString = payloadObject.View().WriteCompact();
+            auto future = m_clientContinuation.Activate(
+                GetModelName(),
+                headers,
+                Crt::ByteBufFromCString(payloadString.c_str()),
+                AWS_EVENT_STREAM_RPC_MESSAGE_TYPE_APPLICATION_MESSAGE,
+                0,
+                onMessageFlushCallback);
+            return future;
+        }
+
+        void ClientOperation::OnContinuationClosed()
+        {
+            /* Because m_initialResponsePromise is a ProtectedPromise, setting its value when it's already
+             * potentially set by `OnContinuationMessage` will just do nothing. */
+            m_initialResponsePromise.SetValue(TaggedResult({EVENT_STREAM_RPC_STREAM_CLOSED_PREMATURELY, 0}));
+
+            m_closedPromise.set_value();
+
+            if (m_streamHandler)
+            {
+                m_streamHandler->OnStreamClosed();
+            }
+        }
+
+        std::future<RpcError> ClientOperation::Close(OnMessageFlushCallback onMessageFlushCallback) noexcept
+        {
+            if (m_isClosed.load())
+            {
+                std::promise<RpcError> alreadyClosedPromise;
+                alreadyClosedPromise.set_value({EVENT_STREAM_RPC_STREAM_CLOSED_PREMATURELY, 0});
+                return alreadyClosedPromise.get_future();
+            }
+            else
+            {
```

Comment:
The protection against closing twice is already accomplished via the underlying continuation. Just use it to send the necessary close message and close of the continuation. If it's already closed, it'll surface via the SendMessage() response